File: src/utils/limits.ts
/**
 * Input-size ceiling applied before anything expensive touches a payload.
 *
 * Every step of the gate is linear in the size of its input — canonical JSON
 * serialisation, SHA-256 hashing, the signal computations — so the gate cannot
 * bound its own latency. A caller (or an attacker) handing it a multi-megabyte
 * argument mapping buys multi-megabyte hashing on the request path.
 *
 * The control therefore has to sit *upstream of scoring*: refuse the call
 * before hashing rather than after. This mirrors the Python line's P5, where
 * the same bound is documented in `docs/gate-properties.md` and the same
 * `MCP_OBSERVATORY_MAX_INPUT_BYTES` variable names the ceiling.
 *
 * Read from the environment on each call rather than cached at module load, so
 * a test or a host can change it without re-importing the module.
 */

export const MAX_INPUT_BYTES_ENV = 'MCP_OBSERVATORY_MAX_INPUT_BYTES';

/** Default ceiling, in bytes. Matches the Python package's default. */
export const DEFAULT_MAX_INPUT_BYTES = 10240;

export function maxInputBytes(): number {
  const raw = process.env[MAX_INPUT_BYTES_ENV];
  if (raw === undefined || raw === '') return DEFAULT_MAX_INPUT_BYTES;
  const parsed = Number(raw);
  // A malformed ceiling must not silently disable the limit.
  if (!Number.isFinite(parsed) || parsed <= 0) return DEFAULT_MAX_INPUT_BYTES;
  return Math.floor(parsed);
}

/** True when `value`'s UTF-8 length exceeds the configured ceiling. */
export function exceedsLimit(value: string | undefined | null): boolean {
  if (value === undefined || value === null) return false;
  return Buffer.byteLength(value, 'utf8') > maxInputBytes();
}

/**
 * True when any argument mapping is over the ceiling.
 *
 * Serialising to measure is itself linear, but it is one pass with no hashing
 * and no signal work behind it, which is the cost the ceiling exists to avoid.
 * A value that cannot be serialised at all (a circular reference, a BigInt) is
 * treated as over the limit: the gate cannot hash what it cannot serialise, so
 * refusing is the fail-secure reading rather than proceeding to a crash.
 */
export function argsExceedLimit(args: unknown): boolean {
  let serialised: string;
  try {
    serialised = JSON.stringify(args) ?? '';
  } catch {
    return true;
  }
  return exceedsLimit(serialised);
}

File: mcp_observatory/utils/limits.py
"""Input-size ceiling shared by the risk-scoring and propose paths.

``docs/gate-properties.md`` P5 measures every scoring step (canonical-JSON
hashing, Jaccard signals, numeric-variance extraction) as linear in input
size, so a gate with no upstream size limit has no bound on its own
latency -- a caller can make the scorer arbitrarily slow simply by sending
an arbitrarily large argument or answer. This module is that limit:
callers check it *before* doing any work proportional to input size, not
after, so an oversized input costs an O(n) length check rather than the
full scoring pipeline.
"""

from __future__ import annotations

import os

from .hashing import canonical_json

MAX_INPUT_BYTES_ENV = "MCP_OBSERVATORY_MAX_INPUT_BYTES"
DEFAULT_MAX_INPUT_BYTES = 10_240


def max_input_bytes() -> int:
    """Resolve the configured ceiling, in bytes of UTF-8 text.

    Falls back to :data:`DEFAULT_MAX_INPUT_BYTES` if the environment
    variable is unset or not a valid integer.
    """
    raw = os.getenv(MAX_INPUT_BYTES_ENV)
    if not raw:
        return DEFAULT_MAX_INPUT_BYTES
    try:
        return int(raw)
    except ValueError:
        return DEFAULT_MAX_INPUT_BYTES


def exceeds_limit(value: str, *, limit: int | None = None) -> bool:
    """True if ``value`` encodes to more UTF-8 bytes than the limit."""
    cap = max_input_bytes() if limit is None else limit
    return len(value.encode("utf-8")) > cap


def any_exceeds_limit(*values: str | None, limit: int | None = None) -> bool:
    """True if any non-``None`` value in ``values`` exceeds the limit."""
    cap = max_input_bytes() if limit is None else limit
    return any(v is not None and exceeds_limit(v, limit=cap) for v in values)


def tool_args_exceed_limit(tool_args, *, limit: int | None = None) -> bool:
    """True if the canonical JSON of ``tool_args`` (as used by ``args_hash``)
    exceeds the byte limit.
    """
    return exceeds_limit(canonical_json(tool_args), limit=limit)
